### ISSUE TITLE
Issues with the Thinkpad Pen Pro

### ISSUE THREAD

**Issue by XrenynTheMage** [linuxwacom/libwacom#198]
I've just bought a Lenovo ThinkPad L13 Yoga (this one: https://pcsupport.lenovo.com/de/de/products/laptops-and-netbooks/thinkpad-l-series-laptops/thinkpad-l13-yoga-type-20r5-20r6?linkTrack=Caps%3ABody_BrowseProduct&searchType=0&keyWordSearch=) with both a touchscreen and a so-called "Thinkpad Pen Pro" for handwriting.
On my Manjaro installation (Arch Linux with gnome) the pen works a little bit, allowing me to move the mouse around and executing right-clicks upon pressing it into the screen really hard. However, Both of the pen's buttons do nothing and I can neither use it to select text nor to draw lines in graphics software such as Krita. It is also not registered by gonme's settings no matter how much I move the pen around, resulting in a picture that looks like this: 
![screenshot from gnome settings -> devices -> Wacom Tablet, cannot be displayed for some reason](http://www.mediafire.com/convkey/3501/wl5g8q6077yrodpzg.jpg)

I have already tried the workaround that solved issue #57 to no effect.

Interestingly though, executing the `$ libwacom-list-local-devices` command from your Troubleshooting page (https://github.com/linuxwacom/libwacom/wiki/Troubleshooting) does nothing for some reason.

I'm on libwacom-1.2-1 and GNOME Version 3.34.2 .

Any suggestions on what else I could try or what additional info I should share?

As much as I love Linux, not being able to take handwritten lecture notes is kind of a big dealbreaker for me right now...

**Comment by whot** [comment]
If `libwacom-list-local-devices` doesn't show anything, the most likely cause is that there is no matching .tablet file for this device. Right now, the best approach is to look at the git history and emulate one of the previous commits to add a tablet data file. The documentation overall is unfortunately a bit lacking, but there's some info here: https://github.com/linuxwacom/libwacom/wiki/Troubleshooting

**Comment by XrenynTheMage** [comment]
Seems legit. One guy on the Lenovo was kind enough to share his tutorial and it worked quite fine for me https://www.reddit.com/r/SurfaceLinux/comments/3mu28a/sp3_pen_tip_button_working/

He isn't manipulating any of the .tablet files though...
Also, the pen still doesn't show up in GNOME's Wacom Tablet settings and its pressure sensitivity is incredibly low for my taste. 
The greatest hurdle for making a custom .tablet file is that there's not really any technical documentation around for the "ThinkPad Pen Pro" that I apparently have, but I think I'm going to give it a shot anyway.

Thanks for the advice [^^]

**Comment by whot** [comment]
worst case if you get it wrong/incomplete we fix it up latter, I wouldn't sweat it.

As for the device not working properly - that's not libwacom's fault since it is basically just a glorified database of text files. To debug that I'll need the output of `libinput record` but please file a separate issue in the xf86-input-wacom component.

**Comment by jigpu** [comment]
Just opened a pull request for the tablet definition at https://github.com/linuxwacom/libwacom/pull/204

**Comment by lsitongia** [comment]
I got one of these laptops about a week ago (model 20R5000VUS) and I am trying to get the button nearest the tip to work. The button furthermost from the tip reports itself as button 2, so I can use xsetwacom to change that to button 3 so that it shows the right-click menu. But, the button near the tip is weird. xev shows it as button 1, which is the same as when the pen taps the screen. xev shows touching and lifting off the screen as button 1 with different states. The button nearest the tip doesn't show anything when it is hovering, but when it is contact with the screen it shows the OPPOSITE states as taping the screen does. In other words, when the pen is in contact with the button nearest the tip is pushed, it generates the same even as lifting the pen off the screen. Then, releasing the button generates the same event as touching the screen with the pen.

My hope was that the button nearest the tip would be a unique number so that I could easily configure it to be middle click (paste text). Why is it weird?

The reason why I'm posting this here is that I found on my laptop that the identifier is slightly different that the one added for this in Github. Instead of 51f6 it is 51f5. So, I created an entry in libwacom for that and now it shows up in Settings for wacom. That is, the tablet shows up but the stylus tab still says "No stylus found". Also, Test Your Settings works, and I can draw in the little window that comes up. So, creating the right libwacom entry didn't change the pen operation. Should it have? 

Is it possible for the pen functions of tapping and the two buttons be buttons 1, 2 and 3?

Thanks.

**Comment by jigpu** [comment]
@lsitongia would you mind following the instructions at the bottom of https://github.com/linuxwacom/wacom-hid-descriptors/ so that we can have additional information about your 51f5 variant? Also, would you mind submitting a pull request for your tablet definition file so that others with the same variant can benefit?

As for the lower-button, rather than being just another button as you were expecting, it is typically wired up as an eraser switch. You should find that when you hold down the button, events start coming out of the "eraser" device rather than the "stylus" device. If you're already in contact with the screen when you press the button, you should see button 1 from the stylus be released, then button 1 from the eraser be pressed (and vice-versa if you release the button while it is touching the screen).

If you don't want it to act as an eraser, you can hack around the behavior with something like `xsetwacom set <eraser id> button 1 2` to have it send a middle-click event when you tap the screen with it held. Unfortunately there's no way to have it send a middle-click while the pen is just hovering since all that does is switch tools rather than send a button event.

**Comment by lsitongia** [comment]
Thank you for your help, @jigpu.
I've run sysinfo.sh and created an issue, as directed in the README.
Do I also need to submit a pull request? Which repository do I do that? In  linuxwacom /
wacom-hid-descriptors ?

**Comment by lsitongia** [comment]
If you don't mind my asking, if you have experience with Ubuntu, can you tell me why the stylus doesn't show up in the Settings panel for Wacom?

**Comment by jigpu** [comment]
I've submitted a pull request for the libwacom tablet definition for you at https://github.com/linuxwacom/libwacom/pull/213

As for why the stylus does not appear in the Settings panel, the "simple" explanation is that libwacom doesn't (yet) have your pen in its stylus definition file. We're working on an update which should add all the known IDs here: https://github.com/linuxwacom/libwacom/pull/178

Unfortunately, the full situation is a bit more tricky: while adding the definition will have the pen show up, GNOME wouldn't properly handle it. Turns out that libwacom and GNOME don't really have good handling for that second "eraser" button, among other things. We're working on the necessary libwacom patches at https://github.com/linuxwacom/libwacom/pull/179 and will be working with GNOME shortly afterwards.

It's going to be a while before fixes appear in the distributions, but we're working on it :)
